### PR TITLE
Swapi

### PR DESCRIPTION
**Pull request by Mrio1**
[Swapi deploy link](https://mrio1.github.io/Incubator/swapi/)

### CODE REVIEW

File: swapi/package.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "swapi",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (maximsan) [inline swapi/package.json:7]: Why scripts are empty ? Please add scripts

File: swapi/src/scripts/dataModel.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+class Model {
+    constructor() {
+		this.resultVolume = 5;
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js:3]: Could you align your code please? Remove an extra spaces on the left side everywhere

File: swapi/src/scripts/dataModel.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    async getApiData(addUrl) {
+    	let apiData = await fetch(`https://swapi.dev/api/${addUrl}`)
+			.then((response) => 
+				response.json()
+			)
+			.catch(()=> {
+				alert("Error. Try refresh page.")
+			})
+    	return apiData
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js:14]: why do you create a variable if you supposed to return result in then ?

File: swapi/src/scripts/dataModel.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		this.resultVolume = 5;
+    }
+
+    async getApiData(addUrl) {
+    	let apiData = await fetch(`https://swapi.dev/api/${addUrl}`)
+			.then((response) => 
+				response.json()
+			)
+			.catch(()=> {
+				alert("Error. Try refresh page.")
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js:12]: please don't use alert for errors. it's annoying, use `console.error`

File: swapi/src/scripts/dataModel.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+			const keys = Object.keys(result).slice(0, this.resultVolume);
+			const newObj = {};
+			keys.forEach(key => {
+				newObj[key] = result[key]
+			})
+			return newObj
+		})
+		return {
+			count: data.count,
+			results: results
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js:34]: use object property value shorthand, please read about it, it's in ES standard from es5 version

File: swapi/src/scripts/dataModel.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+	formatData(data) {
+		const results = data.results.map(result => {
+			const keys = Object.keys(result).slice(0, this.resultVolume);
+			const newObj = {};
+			keys.forEach(key => {
+				newObj[key] = result[key]
+			})
+			return newObj
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js:30]: probably you need to use `reduce` instead. you do a lot of extra actions. also we have `object.assign` and `spread` operator to merge objects..

File: swapi/src/scripts/main.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import "../styles/main.scss";
+import NavigationView from "./navView";
+import Model from "./dataModel";
+import MainFieldView from "./mainFieldView";
+import PaginationView from "./paginationView";
+
+
+//model.getPeopleData(chooseNavCategory)
```
Comment (maximsan) [inline swapi/src/scripts/main.js:8]: remove

File: swapi/src/scripts/main.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    constructor() {
+        this.currentCategory = null;
+        this.currentPage = null;
+        this.isNewCategory = true;
+        this.init();
+    }
+
+    init() {
+        this.navView = new NavigationView(this.chooseNavCategoryHandler.bind(this));
+        this.paginationView = new PaginationView(this.getPaginationCount.bind(this));
```
Comment (maximsan) [inline swapi/src/scripts/main.js:19]: why do you pass action in constructors ? constructor used for properties initialization only

**Review by maximsan — CHANGES_REQUESTED** [review]
I left some common notes. I will review it properly later. Please check and fix them, Thanks!

File: swapi/package.json (hunk 1 above)
Comment (Mrio1) [reply swapi/package.json:7]: I added script/
```suggestion
 "scripts": {
    "start": "webpack"
  }
```

File: swapi/src/scripts/dataModel.js (hunk 2 above)
Comment (Mrio1) [reply swapi/src/scripts/dataModel.js:3]: I deleted extra spaces there and in other places.
```suggestion
class Model {
  constructor() {
    this.resultVolume = 5;
  }
```

File: swapi/src/scripts/dataModel.js (hunk 3 above)
Comment (Mrio1) [reply swapi/src/scripts/dataModel.js:14]: I fixed this.
```suggestion
 async getApiData(addUrl) {
    return await fetch(`https://swapi.dev/api/${addUrl}`)
      .then((response) => response.json())
      .catch((error) => {
        console.error(error);
      });
  }
```

File: swapi/src/scripts/dataModel.js (hunk 4 above)
Comment (Mrio1) [reply swapi/src/scripts/dataModel.js:12]: I fixed this.
```suggestion
.catch((error) => {
        console.error(error);
      });
```

File: swapi/src/scripts/main.js (hunk 7 above)
Comment (Mrio1) [reply swapi/src/scripts/main.js:8]: I removed this.
```suggestion
import PaginationView from "./paginationView";

class Controller {
```

File: swapi/src/scripts/main.js (hunk 8 above)
Comment (Mrio1) [reply swapi/src/scripts/main.js:19]: These actions are callback for ```eventListener```.
```suggestion
class Controller {
  constructor() {
    this.currentCategory = null;
    this.currentPage = null;
    this.isNewCategory = true;
    this.navView = new NavigationView(this.chooseNavCategoryHandler.bind(this));
    this.paginationView = new PaginationView(
      this.getPaginationCount.bind(this)
    );
    this.model = new Model();
    this.mainFieldView = new MainFieldView();
  }
```

File: swapi/src/scripts/dataModel.js [hunk 9]
```diff
@@ -1,6 +1,16 @@
 class Model {
   constructor() {
     this.resultVolume = 5;
+    this.storage = localStorage.getItem('swapi');
+    this.localData = JSON.parse(localStorage.getItem('swapi'));
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js:5]: you have `this.storage` to use

File: swapi/src/scripts/dataModel.js [hunk 10]
```diff
@@ -12,7 +22,13 @@ class Model {
   }
 
   async getData(url, callback) {
-    let data = await this.getApiData(url);
+    let data = this.localData[url];
+    if (!data) {
+      data = await this.getApiData(url);
+      Object.assign(this.localData, {[url]: data});
+      console.log(this.localData)
```
Comment (maximsan) [inline swapi/src/scripts/dataModel.js]: remove

File: swapi/src/scripts/main.js [hunk 11]
```diff
@@ -17,34 +17,60 @@ class Controller {
...
 
-  chooseNavCategoryHandler(category) {
-    if (this.currentCategory !== category) {
+/*   searchByKeyWordHandler(word) {
+    if (this.currentCategory) {
+      this.model.getSearchData(`${this.currentCategory}/?search=${word}`, this.displayCategoryData.bind(this))
+    } else {
+      alert('Choose category!')
+    }
+  } */
```
Comment (maximsan) [inline swapi/src/scripts/main.js]: remove

File: swapi/src/scripts/main.js [hunk 12]
```diff
@@ -17,34 +17,60 @@ class Controller {
...
+
+  chooseNavCategoryHandler(category, searchWord) {
+    if (!category) {
+      if (this.currentCategory) {
+        this.searchWord = searchWord;
+        this.getCategoryData(this.currentCategory, 1);
+        this.currentPage = 1;
+        //this.isNewCategory = true;
```
Comment (maximsan) [inline swapi/src/scripts/main.js]: remove

File: swapi/src/scripts/main.js [hunk 13]
```diff
@@ -17,34 +17,60 @@ class Controller {
...
     this.model.getData(
-      `${this.currentCategory}/${pageUrl}`,
+      `${addUrl}${pageUrl}`,
       this.displayCategoryData.bind(this)
     );
   }
 
-  displayCategoryData({ count, results }) {
+  displayCategoryData(data) {
+    if (!data) { return }
```
Comment (maximsan) [inline swapi/src/scripts/main.js:53]: move return on new line please

**Review by maximsan — APPROVED** [review]
Hello! I left some comment. Please fix them and you can merge it *only* after this. Let me know when you fix them. Thanks!

File: swapi/src/scripts/dataModel.js (hunk 9 above)
Comment (Mrio1) [reply swapi/src/scripts/dataModel.js:5]: I fixed this.
```suggestion
constructor() {
    this.resultVolume = 5;
    this.storage = localStorage.getItem('swapi');
    this.localData = JSON.parse(this.storage);
    if(!this.storage) {
      this.localData = {};
      this.updateLocalData();
    }
```

File: swapi/src/scripts/dataModel.js (hunk 10 above)
Comment (Mrio1) [reply swapi/src/scripts/dataModel.js]: I fixed this.
```
    if (!data) {
      data = await this.getApiData(url);
      Object.assign(this.localData, {[url]: data});
      this.updateLocalData();
    }
```

File: swapi/src/scripts/main.js (hunk 12 above)
Comment (Mrio1) [reply swapi/src/scripts/main.js]: I fixed this.
```suggestion
if (this.currentCategory) {
        this.searchWord = searchWord;
        this.getCategoryData(this.currentCategory, 1);
        this.currentPage = 1;
      }
```

File: swapi/src/scripts/main.js (hunk 13 above)
Comment (Mrio1) [reply swapi/src/scripts/main.js:53]: I fixed this.
```suggestion
displayCategoryData(data) {
    if (!data) { 
      return
    }
```

File: swapi/src/scripts/main.js (hunk 11 above)
Comment (Mrio1) [reply swapi/src/scripts/main.js]: I removed this commented code.

**Pull request merged**